benchmarks/address_manager_store.py

```diff
@@ -0,0 +1,125 @@
+from __future__ import annotations
+
+import asyncio
+import random
+import tempfile
+import time
+from datetime import datetime
+from ipaddress import IPv4Address
+from pathlib import Path
+
+from chia_rs.sized_ints import uint16, uint64
+
+from chia.server.address_manager import (
+    NEW_BUCKETS_PER_ADDRESS,
+    AddressManager,
+    ExtendedPeerInfo,
+)
+from chia.server.address_manager_store import AddressManagerStore
+from chia.types.peer_info import TimestampedPeerInfo
+
+
+def generate_random_ip() -> str:
+    return str(IPv4Address(random.getrandbits(32)))
+
+
+def populate_address_manager(num_new: int = 500000, num_tried: int = 200000) -> AddressManager:
+    random.seed(1337)
+    am = AddressManager()
+    current_time = int(datetime.now().timestamp())
+    total = num_new + num_tried
+
+    for i in range(total):
+        host = generate_random_ip()
+        port = random.randint(1024, 65535)
+        timestamp = current_time - random.randint(0, 100000)
+
+        # Construct TimestampedPeerInfo
+        tpi = TimestampedPeerInfo(host=host, port=uint16(port), timestamp=uint64(timestamp))
+
+        # Create the ExtendedPeerInfo
+        epi = ExtendedPeerInfo(
+            addr=tpi,
+            src_peer=None,  # will default to itself inside constructor
+        )
+
+        am.tried_count += 1  # why do we even have `assert tried_ids != tried_count`?
+        node_id = am.id_count
+        am.id_count += 1
+        epi.random_pos = len(am.random_pos)
+        am.map_info[node_id] = epi
+        am.map_addr[epi.peer_info.host] = node_id
+        am.random_pos.append(node_id)
+
+        if i >= num_new:
+            # make a tried_table entry
+            epi.is_tried = True
+            epi.last_success = timestamp
+            epi.last_try = timestamp - random.randint(0, 1000)
+            bucket = epi.get_tried_bucket(am.key)
+            pos = epi.get_bucket_position(am.key, False, bucket)
+            if am.tried_matrix[bucket][pos] == -1:
+                am.tried_matrix[bucket][pos] = node_id
+                am.tried_count += 1
```

Comment:
is there no function that does this already?
If not, is this code duplicated somewhere in the node as well?
it looks like something that should be a function or two on the address manager

Reply:
The equivalent function in reality would be `AddressManager.add_to_new_table()` but this includes some time consuming operations where it is processed